### PR TITLE
feat(diffusers/loaders): add loaders for controlnet, ti and etc.

### PR DESCRIPTION
**Pull request by sageyou**
# What does this PR do?

add loaders: autoencoders, controlnet, single_file, ip_adapter, textual_inversion.

We provides an easy-to-use LoaderMixin API to load adapter weights

**Textual inversion** is very similar to DreamBooth and it can also personalize a diffusion model to generate certain concepts (styles, objects) from just a few images. Now you can load the textual inversion embeddings with the `load_textual_inversion()` method and generate some images. Let’s load [sd-concepts-library/gta5-artwork](https://huggingface.co/sd-concepts-library/gta5-artwork) embeddings and you’ll need to include the special word <gta5-artwork> in your prompt to trigger it:

```python
from mindone.diffusers import StableDiffusionPipeline
pipeline = StableDiffusionPipeline.from_pretrained("runwayml/stable-diffusion-v1-5") 
pipeline.load_textual_inversion("gta5-artwork/")
prompt = "A cute brown bear eating a slice of pizza, stunning color scheme, masterpiece, illustration, <gta5-artwork> style"
image = pipeline(prompt=prompt)[0][0]
image.save("textual_gta5.png")
```

**IP-Adapter** is a lightweight adapter that enables image prompting for any diffusion model. To start, load a Stable Diffusion checkpoint，then load the IP-Adapter [h94/IP-Adapter](https://huggingface.co/h94/IP-Adapter) weights and add it to the pipeline with the` load_ip_adapter()` method. Once loaded, you can use the pipeline with an image and text prompt to guide the image generation process:

```python
from mindone.diffusers import StableDiffusionPipeline
from mindone.diffusers.utils import load_image
pipeline = StableDiffusionPipeline.from_pretrained("runwayml/stable-diffusion-v1-5")
pipeline.load_ip_adapter("h94/IP-Adapter", subfolder="models", weight_name="ip-adapter_sd15.safetensors")
image=load_image("https://huggingface.co/datasets/huggingface/documentationimages/resolve/main/diffusers/load_neg_embed.png")
image = pipeline(
    prompt='best quality, high quality, wearing sunglasses',
    ip_adapter_image=image,
    negative_prompt="monochrome, lowres, bad anatomy, worst quality, low quality",
    num_inference_steps=50,
)[0][0]
image.save("ipadapter.png")
```

Try it out!

Fixes # (issue)

Adds # (feature)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/mindspore-lab/mindone/blob/master/CONTRIBUTING.md)?
- [ ] Did you make sure to update the documentation with your changes? E.g. record bug fixes or new features in `What's New`. Here are the
      [documentation guidelines](https://github.com/mindspore-lab/mindcv/wiki/%E6%96%87%E6%A1%A3%E7%BC%96%E5%86%99%E8%A7%84%E8%8C%83)
- [ ] Did you build and run the code without any errors?
- [ ] Did you report the running environment (NPU type/MS version) and performance in the doc? (better record it for data loading, model inference, or training tasks)
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@xxx

### CODE REVIEW

File: mindone/diffusers/README.md [hunk 1]
```diff
@@ -117,8 +117,12 @@ Most base, utility and mixin class are available.
 - [x] UniPCMultistepScheduler
 
 ### Loader
-- [ ] LoRA
-- [ ] FromSingleFileMixin
+- [x] LoRA
```
Comment (CaitinZhao) [inline mindone/diffusers/README.md:120]: suggest adding some examples to show how to use and what the effect is

**Review by CaitinZhao — APPROVED** [review]

File: mindone/diffusers/loaders/ip_adapter.py [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+            local_files_only (`bool`, *optional*, defaults to `False`):
+                Whether to only load local model weights and configuration files or not. If set to `True`, the model
+                won't be downloaded from the Hub.
+            token (`str` or *bool*, *optional*):
+                The token to use as HTTP bearer authorization for remote files. If `True`, the token generated from
+                `diffusers-cli login` (stored in `~/.huggingface`) is used.
+            revision (`str`, *optional*, defaults to `"main"`):
+                The specific model version to use. It can be a branch name, a tag name, a commit id, or any identifier
+                allowed by Git.
+            mindspore_dtype (`str` or `mindspore.dtype`, *optional*):
```
Comment (geniuspatrick) [inline mindone/diffusers/loaders/ip_adapter.py:87]: 为什么传入一个`mindspore_dtype`?

File: mindone/diffusers/loaders/ip_adapter.py [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+                        if image_encoder_folder.count("/") == 0:
+                            image_encoder_subfolder = Path(subfolder, image_encoder_folder).as_posix()
+                        else:
+                            image_encoder_subfolder = Path(image_encoder_folder).as_posix()
+
+                        image_encoder = CLIPVisionModelWithProjection.from_pretrained(
+                            pretrained_model_name_or_path_or_dict,
+                            subfolder=image_encoder_subfolder,
+                        )
+                        if mindspore_dtype is not None:
```
Comment (geniuspatrick) [inline mindone/diffusers/loaders/ip_adapter.py:172]: 没有保持一致

File: mindone/diffusers/loaders/ip_adapter.py [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+                    )
+
+            # create feature extractor if it has not been registered to the pipeline yet
+            if hasattr(self, "feature_extractor") and getattr(self, "feature_extractor", None) is None:
+                feature_extractor = CLIPImageProcessor()
+                self.register_modules(feature_extractor=feature_extractor)
+
+        # load ip-adapter into unet
+        unet = getattr(self, self.unet_name) if not hasattr(self, "unet") else self.unet
+        print("before unet load ip adapter into unet")
```
Comment (geniuspatrick) [inline mindone/diffusers/loaders/ip_adapter.py:193]: 去掉print

File: mindone/diffusers/loaders/single_file.py [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+    fetch_ldm_config_and_checkpoint,
+    infer_model_type,
+)
+
+logger = logging.get_logger(__name__)
+
+# Pipelines that support the SDXL Refiner checkpoint
+REFINER_PIPELINES = [
+    "StableDiffusionXLImg2ImgPipeline",
+    "StableDiffusionXLInpaintPipeline",
```
Comment (geniuspatrick) [inline mindone/diffusers/loaders/single_file.py:33]: control可以加上

File: mindone/diffusers/loaders/single_file.py [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+                init_kwargs.update(components)
+
+        additional_components = set_additional_components(
+            class_name, original_config, checkpoint=checkpoint, model_type=model_type
+        )
+        if additional_components:
+            init_kwargs.update(additional_components)
+
+        init_kwargs.update(passed_pipe_kwargs)
+        pipe = pipeline_class(**init_kwargs)
```
Comment (geniuspatrick) [inline mindone/diffusers/loaders/single_file.py:304]: `pipe.to`怎么删掉了

File: mindone/diffusers/loaders/single_file_utils.py [hunk 7]
```diff
@@ -0,0 +1,1547 @@
...
+STABLE_CASCADE_DEFAULT_CONFIGS = {
+    "stage_c": {"pretrained_model_name_or_path": "diffusers/stable-cascade-configs", "subfolder": "prior"},
+    "stage_c_lite": {"pretrained_model_name_or_path": "diffusers/stable-cascade-configs", "subfolder": "prior_lite"},
+    "stage_b": {"pretrained_model_name_or_path": "diffusers/stable-cascade-configs", "subfolder": "decoder"},
+    "stage_b_lite": {"pretrained_model_name_or_path": "diffusers/stable-cascade-configs", "subfolder": "decoder_lite"},
+}
+
+
+def convert_stable_cascade_unet_single_file_to_diffusers(original_state_dict):
+    is_stage_c = "clip_txt_mapper.weight" in original_state_dict
```
Comment (geniuspatrick) [inline mindone/diffusers/loaders/single_file_utils.py:86]: 这块的差异好像很大

File: mindone/diffusers/loaders/textual_inversion.py [hunk 8]
```diff
@@ -0,0 +1,518 @@
...
+# limitations under the License.
+from typing import Dict, List, Optional, Union
+
+from huggingface_hub.utils import validate_hf_hub_args
+from transformers import PreTrainedModel, PreTrainedTokenizer
+
+import mindspore
+from mindspore import nn, ops
+
+from ..models.modeling_utils import load_state_dict
```
Comment (geniuspatrick) [inline mindone/diffusers/loaders/textual_inversion.py:22]: 为什么要用这个函数？

File: mindone/diffusers/loaders/textual_inversion.py [hunk 9]
```diff
@@ -0,0 +1,518 @@
...
+
+    @staticmethod
+    def _retrieve_tokens_and_embeddings(tokens, state_dicts, tokenizer):
+        all_tokens = []
+        all_embeddings = []
+        for state_dict, token in zip(state_dicts, tokens):
+            if isinstance(state_dict, mindspore.Tensor):
+                if token is None:
+                    raise ValueError(
+                        "You are trying to load a textual inversion embedding that has been saved as a Mindspore tensor. \
```
Comment (geniuspatrick) [inline mindone/diffusers/loaders/textual_inversion.py:200]: 不要使用反斜杠

File: mindone/diffusers/loaders/unet.py [hunk 10]
```diff
@@ -356,3 +368,241 @@ def delete_adapters(self, adapter_names: Union[List[str], str]):
...
+        # convert IP-Adapter Image Projection layers to diffusers
+        image_projection_layers = []
+        for state_dict in state_dicts:
+            image_projection_layer = self._convert_ip_adapter_image_proj_to_diffusers(state_dict["image_proj"])
+            image_projection_layers.append(image_projection_layer)
+
+        self.encoder_hid_proj = MultiIPAdapterImageProjection(image_projection_layers)
+        self.config.encoder_hid_dim_type = "ip_image_proj"
+        self.encoder_hid_dim_type = "ip_image_proj"
+        self.to(dtype=self.dtype)
```
Comment (geniuspatrick) [inline mindone/diffusers/loaders/unet.py:517]: `encoder_hid_dim_type`

File: mindone/transformers/modeling_ms_utils.py [hunk 11]
```diff
@@ -1126,3 +1126,220 @@ def convert_state_dict(m, state_dict_pt):
         )
         mismatched_keys, offload_index, error_msgs = [], 0, ""
         return model, missing_keys, unexpected_keys, mismatched_keys, offload_index, error_msgs
+
+    def resize_token_embeddings(
```
Comment (geniuspatrick) [inline mindone/transformers/modeling_ms_utils.py:1130]: rebase之后再看一下

File: mindone/diffusers/README.md [hunk 12]
```diff
@@ -106,8 +106,12 @@ Most base, utility and mixin class are available.
 - [ ] DPMSolverSDEScheduler
 
 ### Loader
```
Comment (geniuspatrick) [inline mindone/diffusers/README.md:108]: all supported

File: mindone/diffusers/loaders/ip_adapter.py (hunk 3 above)
Comment (sageyou) [reply mindone/diffusers/loaders/ip_adapter.py:172]: diffusers 这里创建image_encoder的时候代码是：
```python
image_encoder = CLIPVisionModelWithProjection.from_pretrained(
    pretrained_model_name_or_path_or_dict,
    subfolder=image_encoder_subfolder,
    low_cpu_mem_usage=low_cpu_mem_usage,
).to(self.device, dtype=self.dtype)
```
我们的piepline 没有`dtype`这个属性。 当pipeline的dtype是float16时，下面代码会有数据类型不一致的报错：
```python
pipeline = StableDiffusionPipeline.from_pretrained(
    "runwayml/stable-diffusion-v1-5",       
    mindspore_dtype=mindspore.float16
)
pipeline.load_ip_adapter("IP-Adapter", subfolder="models", weight_name="ip-adapter_sd15.safetensors",）
```
所以我在这里传入了mindspore_dtype 这个参数。
但是或许有更好的idea？

**Review by geniuspatrick — APPROVED** [review]

**Pull request merged**